After upgrade to from 2.3.1 to 2.4.2, NPE in Whitelist.<init>

We upgraded and did not change our config, now we are getting an NPE:

Config is (with some internal values elided):

```
primary-meta-store:
  access-control-type: READ_AND_WRITE_ON_DATABASE_WHITELIST
  database-prefix: ''
  name: primary
  remote-meta-store-uris: thrift://<removed>:9083
  writable-database-white-list:
  - <removed>_.*
  - <removed>_.*

federated-meta-stores:

  - name: <removed>
    database-prefix: <removed>_
    remote-meta-store-uris: thrift://<removed>:9083
    mapped-databases:
```
NPE is:

```
2018-08-30 16:11:25.367 ERROR 1 --- [pool-4-thread-1] o.a.t.s.TThreadPoolServer : Error occurred during processing of message.
java.lang.RuntimeException: Error creating TProcessor
at com.hotels.bdp.waggledance.server.TSetIpAddressProcessorFactory.getProcessor(TSetIpAddressProcessorFactory.java:56) ~[classes!/:?]
at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:268) [libthrift-0.9.3.jar!/:0.9.3]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_181]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_181]
at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
Caused by: java.lang.NullPointerException
at com.hotels.bdp.waggledance.util.Whitelist.<init>(Whitelist.java:31) ~[classes!/:?]
at com.hotels.bdp.waggledance.mapping.service.impl.PrefixBasedDatabaseMappingService.getWhitelistedDatabases(PrefixBasedDatabaseMappingService.java:105) ~[classes!/:?]
at com.hotels.bdp.waggledance.mapping.service.impl.PrefixBasedDatabaseMappingService.add(PrefixBasedDatabaseMappingService.java:98) ~[classes!/:?]
at com.hotels.bdp.waggledance.mapping.service.impl.PrefixBasedDatabaseMappingService.init(PrefixBasedDatabaseMappingService.java:85) ~[classes!/:?]
at com.hotels.bdp.waggledance.mapping.service.impl.PrefixBasedDatabaseMappingService.<init>(PrefixBasedDatabaseMappingService.java:78) ~[classes!/:?]
at com.hotels.bdp.waggledance.server.FederatedHMSHandlerFactory.createDatabaseMappingService(FederatedHMSHandlerFactory.java:72) ~[classes!/:?]
at com.hotels.bdp.waggledance.server.FederatedHMSHandlerFactory.create(FederatedHMSHandlerFactory.java:56) ~[classes!/:?]
at com.hotels.bdp.waggledance.server.TSetIpAddressProcessorFactory.getProcessor(TSetIpAddressProcessorFactory.java:50) ~[classes!/:?]
... 4 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,7 @@
 ## TBD
 ### Fixed
 * Support for regex `mapped-databases` for MANUAL database resolution. See [#147](https://github.com/HotelsDotCom/waggle-dance/issues/147).
+* Avoid NPE when no elements are provided for `mapped-databases` in the configuration. See [#131](https://github.com/HotelsDotCom/waggle-dance/issues/131).
 
 ## [3.1.2] - 2019-01-11
 ### Changed

--- a/waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/FederatedMetaStore.java
+++ b/waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/FederatedMetaStore.java
@@ -20,6 +20,9 @@ import java.util.List;
 
 import org.hibernate.validator.constraints.NotBlank;
 
+import com.hotels.bdp.waggledance.api.ValidationError;
+import com.hotels.bdp.waggledance.api.WaggleDanceException;
+
 public class FederatedMetaStore extends AbstractMetaStore {
 
   private List<String> mappedDatabases = Collections.emptyList();
@@ -67,6 +70,10 @@ public class FederatedMetaStore extends AbstractMetaStore {
   }
 
   public void setMappedDatabases(List<String> mappedDatabases) {
+    if (mappedDatabases == null) {
+      ValidationError validationError = ValidationError.builder().error("mapped-databases should not be null").build();
+      throw new WaggleDanceException(validationError.getErrorMessage());
+    }
     this.mappedDatabases = mappedDatabases;
   }
 }

--- a/waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/FederatedMetaStore.java
+++ b/waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/FederatedMetaStore.java
@@ -18,14 +18,13 @@ package com.hotels.bdp.waggledance.api.model;
 import java.util.Collections;
 import java.util.List;
 
-import org.hibernate.validator.constraints.NotBlank;
+import javax.validation.constraints.NotNull;
 
-import com.hotels.bdp.waggledance.api.ValidationError;
-import com.hotels.bdp.waggledance.api.WaggleDanceException;
+import org.hibernate.validator.constraints.NotBlank;
 
 public class FederatedMetaStore extends AbstractMetaStore {
 
-  private List<String> mappedDatabases = Collections.emptyList();
+  private @NotNull List<String> mappedDatabases = Collections.emptyList();
 
   public FederatedMetaStore() {}
 
@@ -70,10 +69,6 @@ public class FederatedMetaStore extends AbstractMetaStore {
   }
 
   public void setMappedDatabases(List<String> mappedDatabases) {
-    if (mappedDatabases == null) {
-      ValidationError validationError = ValidationError.builder().error("mapped-databases should not be null").build();
-      throw new WaggleDanceException(validationError.getErrorMessage());
-    }
     this.mappedDatabases = mappedDatabases;
   }
 }

--- a/waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/FederatedMetaStoreTest.java
+++ b/waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/FederatedMetaStoreTest.java
@@ -29,8 +29,6 @@ import org.junit.Test;
 import com.fasterxml.jackson.databind.MapperFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
-import com.hotels.bdp.waggledance.api.WaggleDanceException;
-
 public class FederatedMetaStoreTest extends AbstractMetaStoreTest<FederatedMetaStore> {
 
   public FederatedMetaStoreTest() {
@@ -91,9 +89,11 @@ public class FederatedMetaStoreTest extends AbstractMetaStoreTest<FederatedMetaS
     assertThat(metaStore.getMappedDatabases(), is(mappedDatabases));
   }
 
-  @Test(expected = WaggleDanceException.class)
+  @Test
   public void nullMappedDatabases() {
     metaStore.setMappedDatabases(null);
+    Set<ConstraintViolation<FederatedMetaStore>> violations = validator.validate(metaStore);
+    assertThat(violations.size(), is(1));
   }
 
   @Test

--- a/waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/FederatedMetaStoreTest.java
+++ b/waggle-dance-api/src/test/java/com/hotels/bdp/waggledance/api/model/FederatedMetaStoreTest.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2016-2018 Expedia Inc.
+ * Copyright (C) 2016-2019 Expedia Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -28,6 +28,8 @@ import org.junit.Test;
 
 import com.fasterxml.jackson.databind.MapperFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
+
+import com.hotels.bdp.waggledance.api.WaggleDanceException;
 
 public class FederatedMetaStoreTest extends AbstractMetaStoreTest<FederatedMetaStore> {
 
@@ -87,6 +89,11 @@ public class FederatedMetaStoreTest extends AbstractMetaStoreTest<FederatedMetaS
     mappedDatabases.add("database");
     metaStore.setMappedDatabases(mappedDatabases);
     assertThat(metaStore.getMappedDatabases(), is(mappedDatabases));
+  }
+
+  @Test(expected = WaggleDanceException.class)
+  public void nullMappedDatabases() {
+    metaStore.setMappedDatabases(null);
   }
 
   @Test

